src/topology/mod.rs

```diff
@@ -118,24 +118,27 @@ impl RunningTopology {
         let mut running_tasks = self.tasks;
 
         let mut wait_handles = Vec::new();
-        let mut check_handles = HashMap::new();
+        let mut check_handles = HashMap::<String, Vec<_>>::new();
 
-        for (name, task) in running_tasks.drain() {
+        for (name, task) in running_tasks.drain().chain(self.source_tasks.into_iter()) {
```

Comment:
Could you explain this change? I'm curious why source tasks aren't included in `running_tasks` already and why we now need to track multiple handles per task.

Reply:
Also I'd like to see some comments explaining some of this logic if we make it a bit more complex its starting to get harder to follow 😄

Reply:
As far as I can see, `source_tasks` is needed because we treat their tasks differently. For example when we are shutting down sources we are [waiting](https://github.com/timberio/vector/blob/1c01af7bd1ac7fd0b27e5517907480995104f328/src/topology/mod.rs#L339) on their tasks, while we are just [forgetting](https://github.com/timberio/vector/blob/1c01af7bd1ac7fd0b27e5517907480995104f328/src/topology/mod.rs#L349) other ones. 

With this change we stop dropping their tasks, and instead we started treating them as regular tasks. This is necessary so to give them time to shutdown gracefully through `ShutdownSignal`. 

Regarding multiple handles per task, we actually aren't doing that, there is only one handle per task, but we have two tasks under the same name, [one](https://github.com/timberio/vector/blob/1c01af7bd1ac7fd0b27e5517907480995104f328/src/topology/builder.rs#L154) for `source`  and [the other](https://github.com/timberio/vector/blob/1c01af7bd1ac7fd0b27e5517907480995104f328/src/topology/builder.rs#L153) for `pump`, hence the need for `Vec`.